### DISCUSSION TITLE
Mapping over another operator doesn't work with error "PythonOperator.partial() got unexpected keyword arguments"

### DISCUSSION THREAD

**Discussion by shutongy** [apache/airflow#31974] [Q&A]
Hi all, I'm trying to use the "mapping over another operator" functionality documented [here](https://docs.astronomer.io/learn/dynamic-tasks?tab=two-traditional#mapping-over-the-result-of-another-operator), and my implementation is exactly the same as the code example; but it gives me an error "PythonOperator.partial() got unexpected keyword arguments" . Our AF is on version 2.3.3.

I think the issue might be due to how we subclass DAG, but I'm not 100% sure, hence wanna double check here..

We subclassed the DAG class, call it MyOwnDAG, a simplified implementation is like this:
```
class MyOwnDAG(DAG):
    def __init__(self, ...) -> DAG:
        super().__init__(
            dag_id=dag_id,
            ...  # omitted
            *args,
            **kwargs)
      
        self.default_args = {
        	'foo': 1
               'bar': 2
        }
```

All our DAGs uses this subclass, instead of the built-in DAG class.

I read through the implementation of these two files

- [BaseOperator](https://airflow.apache.org/docs/apache-airflow/stable/_modules/airflow/models/baseoperator.html)

- [MappedOperator](https://airflow.apache.org/docs/apache-airflow/2.4.0//_modules/airflow/models/mappedoperator.html)

My GUESS is that this is the function call stack:

`BaseOperator.partial()` --> `validate_mapping_kwargs` --> `param_names = init._BaseOperatorMeta__param_names` -?-> `BaseOperatorMeta. __new__` --> `BaseOperatorMeta._apply_defaults` -?-> `MyOwnDAG.__init__`, which then adds `foo: bar` to the `default_args` dictionary, and those keys are recognized, hence `validate_mapping_kwargs` fails the validation.

**OR**, another possibility is that our AF cluster is configured to always send some specific op_args into every operator call, which then makes it fail the validation, but it's a little hard to believe this is the cause -- it should be very common for an AF cluster to configure some default args for all operators.

Does that ^ make sense? And if that's the case, I wonder how should I resolve this? Thanks!

**Comment by potiuk** [comment]
Don't subclass DAGs.  There are not supposed to be subclassed.

**Comment by carloscdias** [comment]
@potiuk I recently came across this issue, could you please point out reasons/docs for not subclassing DAGs?
Thanks!

**Reply by potiuk** [reply]
Could you point the reason why you would like to subclass them ?

**Reply by carloscdias** [reply]
Actually no, but I could not find this recommendation anywhere and I suppose that other people could not find it either, which led me to come across this issue. Now that it is already done and spread in the sense that reverting this subclass would introduce major issues, I was think on:
1. What can be done to attenuate issues that could arise from subclassing DAG, such as this one
2. To find a documentation to point out in future events

**Reply by potiuk** [reply]
Well the problem is that in Python you can extend ANY class and can do ANYTHING with it. This is unlike in some languages  (Java) where you have encapsulation, final etc. etc. Python has all the dunder methods that implement different behaviours, metclasses and many other things. So when you extend class that comes from the `dag.py` fila and has > 4000 lines, which has no explanations "You can extent this class in this or that way", well If I were you I'd be extremely careful to understand what I am doing myself by careful reading of that code.

This is general advice, not specific to DAG.  It's rather impossible to say what you can or cannot do there if you ask me. Or maybe possible but people who contribute here are volunteers and you seem to expect they will be able to help you to not have to understand the code you are extending and spend their time on creative ways users might implement some things. Or maybe I am misinterpreting things?

My point is simply - only classes that you understand or that were foreseen for extension and documented should be extended in Python. Quite the opposite of your understanding "I want to extend huge class and I expect documentation describing If I can /cannot do and how". I think it's a bit backwards. We should start with "why would you want to extend DAG and this the way how you can do it". And for that I am super-open to get contributions on.

Explain how you want to extend DAG and way and if we find it's a good idee - you will follow it then with a PR starting a chapter "Those are the good ways you can extend DAG to achieve this".  This would be a fantastic contribution that could help others like you who want to extend DAG.

I would very much like this kind of documentation and I thin adding would be a great idea and, I am more than happy to review such proposals and such PRs.

**Reply by potiuk** [reply]
BTW. The good thing about those kind of explanations and PRs are also that 

a) you might find you do not need to extend the class at all, maybe there are ways this can be done already
b) maybe it's a good idea to find a way into the DAG class itself as contribution.

So let me repeat the question (and possibly it's a good idea to split it out to a separate discussion) - what do you want to achieve by extending DAG ? What did you do to achieve that? Why other things did not work for you? And eventually we will see if that can be contributed back as documentation or code and we will kindly ask you to do so.

**Reply by carloscdias** [reply]
Thank you for your reply. I appreciate your insights regarding the extensibility of Python classes. Just to clarify, my main goal is to address an existing problem within the codebase, not to extend its behavior, but to fix an earlier mistake or misdirection that someone wanting to extend the DAG behavior may have done. I'm not seeking explanations or modifications but rather focusing on fixing the issue at hand.

You may also be interested to know that Python provides the "final" decorator to restrict class inheritance, which can be useful in controlling extensions. You can find more information on the "final" decorator in the [Python documentation](https://docs.python.org/3/library/typing.html#typing.final).

My specific concern is to understand what a class extending DAG, while adding properties, might be missing that compromises other operators, such as "partial." This insight will help in rectifying the issue and improving the codebase.

Once I've identified and resolved the problem, I'll be more than happy to contribute to this repository by writing documentation about the issue and its solution. Thank you for your input, and I hope this update clearly conveys my intention and willingness to assist.

**Reply by potiuk** [reply]
> You may also be interested to know that Python provides the "final" decorator to restrict class inheritance, which can be useful in controlling extensions. You can find more information on the "final" decorator in the [Python documentation](https://docs.python.org/3/library/typing.html#typing.final).

You may also be interested @carloscdias  that this is what one of the maintainers looking at the discussion here tried and failed as we were discussing it https://github.com/apache/airflow/pull/35360 - also because this class IS internally extended already in a specific way.

> My specific concern is to understand what a class extending DAG, while adding properties, might be missing that compromises other operators, such as "partial." This insight will help in rectifying the issue and improving the codebase.

Absolutely it's also my concetn.  Read the code and understand that and try to solve the problem and come back with PR. This is precisely what I proposed. DAG class was written by likely 100s of people over the course of last almost 10 years, 4000+ lines of codes, 150 methods and a bunch of other classes using it.

Probably there is not a single person who knows exactly "how to add a new property" - anyone doing it would have to do exactly this - go to the code, read and understand it, solve the problems, run tests etc if there is a legitimate need to extend the class. And it would be really, really fantastic to describe it in a comprehensive way as a use case. So if you want to do it - absolutely go for it.

This is all what I am telling - if you are not ready for that task - don't "expect" this to be described - because maintainers have not foreseen that as the way DAG should be extended by the users. But if you want to make it so - the best way is to try it and contribute back documentation about it.

**Reply by potiuk** [reply]
BTW. @carloscdias  I hope the whole discussion is precisely answering your question:

> @potiuk I recently came across this issue, could you please point out reasons/docs for not subclassing DAGs?
Thanks!

All the words written above and time spent on discussing it take a lot more time than writing this:

> Don't subclass DAGs. There are not supposed to be subclassed.

I did not write it "Cannot" be extended, I wrote it's not "Supposed" to be. And advised the user not to do it.
So while my initial answer's brevity could be misleading it was just short way of expressing all the above. 

* extension of DAG is not "foreseen" by maintainers
* if you do it - understand what you are doing 
* and when you find a good reason for it, and good "ways" how it can be extended (and made "extendable") - come back and contribute it and make DAG as "foreseen" to be extended and documented - It's not currently.

**Reply by carloscdias** [reply]
It does. Thanks!
